src/lib/krb5/asn.1/asn1_encode.c

```diff
@@ -26,105 +26,100 @@
 
 #include "asn1_encode.h"
 
+struct asn1buf_st {
+    uint8_t *ptr;               /* Position, moving backwards; may be NULL */
+    size_t count;               /* Count of bytes written so far */
+};
+
 /**** Functions for encoding primitive types ****/
 
-krb5_error_code
-k5_asn1_encode_bool(asn1buf *buf, intmax_t val, size_t *len_out)
+/* Insert one byte into buf going backwards. */
+static inline void
+insert_byte(asn1buf *buf, uint8_t o)
+{
+    if (buf->ptr != NULL)
+        *--buf->ptr = o;
+    buf->count++;
+}
+
+/* Insert a block of bytes into buf going backwards (but without reversing
+ * bytes). */
+static inline void
+insert_bytes(asn1buf *buf, const void *bytes, size_t len)
 {
-    uint8_t bval = val ? 0xFF : 0x00;
+    if (buf->ptr != NULL) {
+        memcpy(buf->ptr - len, bytes, len);
+        buf->ptr -= len;
+    }
+    buf->count += len;
+}
 
-    *len_out = 1;
-    return asn1buf_insert_octet(buf, bval);
+void
+k5_asn1_encode_bool(asn1buf *buf, intmax_t val)
+{
+    insert_byte(buf, val ? 0xFF : 0x00);
 }
 
-krb5_error_code
-k5_asn1_encode_int(asn1buf *buf, intmax_t val, size_t *len_out)
+void
+k5_asn1_encode_int(asn1buf *buf, intmax_t val)
 {
-    krb5_error_code ret;
-    size_t len = 0;
     long valcopy;
     int digit;
 
     valcopy = val;
     do {
         digit = valcopy & 0xFF;
-        ret = asn1buf_insert_octet(buf, digit);
-        if (ret)
-            return ret;
-        len++;
+        insert_byte(buf, digit);
```

Comment:
`digit` is an `int` at the time of this call, but the function takes a `uint8_t`.  I think usually we eschew this kind of cast, but it's being used fairly often in this PR.

Reply:
I don't mind making `digit` uint8_t in this function and k5_asn1_encode_uint(), but in general I'm okay with implicit downcasting as long as we have reason to believe the value will be in the range of the target integer type, as is the case for all of the insert_byte() calls.

Reply:
Thanks for clarifying.